Bmx.Service.Aws/AwsClient.cs

```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web;
+using System.Xml;
+using Amazon.SecurityToken;
+using Amazon.SecurityToken.Model;
+using Bmx.Core;
+using Bmx.Service.Aws.Models;
+
+namespace Bmx.Service.Aws {
+	public class AwsClient : ICloudProvider {
+		private readonly AmazonSecurityTokenServiceClient _stsClient;
+
+		private string _samlString;
+		private XmlDocument _samlToken;
+		private List<AwsRole> _awsRoles;
+
+		public AwsClient() {
+			_stsClient = new AmazonSecurityTokenServiceClient();
+		}
+
+		public void SetSamlToken( string encodedSaml ) {
+			_samlString = encodedSaml;
+
+			var samlStatements = _samlString.Split( ";" );
+			// Process the B64 Encoded SAML string to get valid XML doc
+			var samlString = new StringBuilder();
+			foreach( var inputValueString in samlStatements ) {
+				samlString.Append( HttpUtility.HtmlDecode( inputValueString ) );
+			}
+
+			var samlData = Convert.FromBase64String( samlString.ToString() );
+
+			_samlToken = new XmlDocument();
+			_samlToken.LoadXml( Encoding.UTF8.GetString( samlData ) );
+		}
+
+		public string[] GetRoles() {
+			var roleNodes = _samlToken.SelectNodes( "//*[@Name=\"https://aws.amazon.com/SAML/Attributes/Role\"]/*" );
+
+			_awsRoles = new List<AwsRole>();
+
+			foreach( XmlElement roleNode in roleNodes ) {
+				// SAML has value: <principal-arn>, <role-arn>
+				// The last part of the role-arn is a human readable name
+				var nodeContents = roleNode.InnerText.Split( "," );
+
+				_awsRoles.Add( new AwsRole {
+					PrincipalArn = nodeContents[0],
+					RoleArn = nodeContents[1],
+					RoleName = nodeContents[1].Split( "/" )[1]
+				} );
+			}
+
+			return _awsRoles.Select( role => role.RoleName ).ToArray();
+		}
```

Comment:
The way `_samlToken` and `_awsRoles` is set suggests that we should use a builder pattern maybe? Something something

```cs
class AwsClient {
  public AwsClient ( AwsRoles[] awsRoles ) {
    _awsRoles = awsRoles
    _stsClient = new AmazonSecurityTokenServiceClient();
  }
  ...
  static AwsClient Create( string encodedSaml ) {
    ...
    var samlToken = LoadXml( ... );
    ...
    var awsRoles = ....
    ...
    return new AwsClient( awsRoles );
  }
}
```
?

This will prevent callers from calling `GetRoles()` or `GetTokens()` before the internals are fully set up.

Reply:
Thanks! I really didn't like how I had it setup, this pattern solves that problem very well.

Reply:
While trying to implement DI and Unit Tests I've realized this issue extends deeper than just this setting. 

Services for Okta and AWS both are stateful, and there is a very strong temporal coupling (methods needing to be called in a very specific order for the app to work without runtime errors). 

As far as I can see there is no way to remove this coupling without making the services stateless (and by leaking internal implementation details from those classes into BmxCore).

tl;dr its an extensive issue that goes deeper than just having the presence encodedSaml guaranteed via builder pattern, requires investigation on its own IMO.

Reply:
Oof gotcha. Makes sense to me.